contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,660 @@
+                                        total_liquidity: u128|
+             -> Result<u128, DexError> {
+                let op1 = withdraw_amount
+                    .checked_div(token_balance)
+                    .ok_or(DexError::Arithmethic)?;
+
+                let op2 = 1u128.checked_sub(op1).ok_or(DexError::Arithmethic)?;
+
+                let op3 = op2.nth_root(4);
+
+                let op4 = 1u128.checked_sub(op3).ok_or(DexError::Arithmethic)?;
+
+                let op5 = total_liquidity
+                    .checked_mul(op4)
+                    .ok_or(DexError::Arithmethic)?;
+
+                Ok(op5)
+            };
+
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+
+            // check role, only designated account can add liquidity
+            <Self as AccessControlled>::check_role(
+                self.access_control,
+                caller,
+                Role::LiquidityProvider(this),
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            )?;
+
+            let native_balance = self.env().balance();
+            let total_liquidity = self.total_liquidity;
+            let redeemed_liquidity = pool_shares_redeemed(amount, native_balance, total_liquidity)?;
+
+            // this should return Err variant if user has not enough liquidity so we can omit an explicit check
+            let caller_liquidity = self
+                .liquidity
+                .get(&caller)
+                .unwrap_or(0)
+                .checked_sub(redeemed_liquidity)
+                .ok_or(DexError::Arithmethic)?;
+
+            self.liquidity.insert(&caller, &caller_liquidity);
+
+            self.total_liquidity = self
+                .total_liquidity
+                .checked_sub(redeemed_liquidity)
+                .ok_or(DexError::Arithmethic)?;
+
+            self.env()
+                .transfer(caller, amount)
+                .map_err(|why| DexError::NativeTransferFailed(format!("{:?}", why)))?;
+
+            // emit event
+            Self::emit_event(
+                self.env(),
+                Event::NativeWithdrawal(NativeWithdrawal {
+                    caller,
+                    amount,
+                    redeemed_liquidity,
+                }),
+            );
+
+            Ok(())
+        }
+
+        /// Alters the swap_fee parameter
+        ///
+        /// Can only be called by the contract's Admin.
+        #[ink(message)]
+        pub fn set_swap_fee(&mut self, swap_fee: Balance) -> Result<(), DexError> {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+
+            <Self as AccessControlled>::check_role(
+                self.access_control,
+                caller,
+                Role::Admin(this),
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            )?;
```

Comment:
I would really love to work towards macro-ing / hiding access control; the solution I would appreciate would be like

```rust
#[ink(message)]
#[access_control(require = Admin)]
pub fn set_swap_fee(&mut self, swap_fee: Balance) -> Result<(), DexError> {
  // logic that works under assumption that the caller has all required roles
```

such macro in standard Rust env would be really easy to implement (I can volutneer :P) but I'm not so sure how they work with ink!